File: tools/ocr/page_crop.py
#!/usr/bin/env python3
"""Crop cookbook page photos before OCR.

This script is designed for the Linux/GPU OCR pipeline. It accepts the same
contract as the macOS Swift cropper:

    page_crop.py [--metadata PATH] [--inset RATIO] INPUT_IMAGE OUTPUT_PNG

It first tries an OpenCV contour/perspective crop when cv2 is installed. If
OpenCV is unavailable, or no strong quadrilateral is found, it falls back to a
NumPy page-background crop that removes neighboring-page spill without requiring
Apple Vision.
"""

from __future__ import annotations

import argparse
import json
import math
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

try:
    import numpy as np
except Exception as exc:  # pragma: no cover - dependency failure path.
    print(f"could not import numpy: {exc}", file=sys.stderr)
    raise SystemExit(3)

try:  # Pillow is the lightweight image I/O fallback.
    from PIL import Image, ImageOps
except Exception:  # pragma: no cover - exercised only in alternate envs.
    Image = None
    ImageOps = None

try:  # OpenCV is optional but gives better perspective correction.
    import cv2
except Exception:  # pragma: no cover - local test env may not have cv2.
    cv2 = None


@dataclass
class CropResult:
    image: np.ndarray
    method: str
    area_ratio: float
    points: list[tuple[float, float]]
    confidence: float | None = None


class CropError(Exception):
    pass


def point_report(point: tuple[float, float]) -> dict[str, float]:
    return {"x": float(point[0]), "y": float(point[1])}


def load_rgb(path: Path) -> np.ndarray:
    if Image is not None and ImageOps is not None:
        with Image.open(path) as image:
            return np.asarray(ImageOps.exif_transpose(image).convert("RGB"))

    if cv2 is not None:
        bgr = cv2.imread(str(path), cv2.IMREAD_COLOR)
        if bgr is None:
            raise CropError(f"could not load image: {path}")
        return cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)

    raise CropError("could not import Pillow or OpenCV for image I/O")


def save_png(image: np.ndarray, path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    output = np.clip(image, 0, 255).astype(np.uint8)

    if Image is not None:
        Image.fromarray(output, mode="RGB").save(path, "PNG")
        return

    if cv2 is not None:
        ok = cv2.imwrite(str(path), cv2.cvtColor(output, cv2.COLOR_RGB2BGR))
        if ok:
            return

    raise CropError(f"could not write image: {path}")


def resize_rgb(image: np.ndarray, max_dimension: int) -> tuple[np.ndarray, float]:
    height, width = image.shape[:2]
    scale = min(1.0, max_dimension / max(width, height))
    if scale >= 1:
        return image, 1.0

    new_size = (max(1, int(round(width * scale))), max(1, int(round(height * scale))))
    if cv2 is not None:
        return cv2.resize(image, new_size, interpolation=cv2.INTER_AREA), scale
    if Image is not None:
        resized = Image.fromarray(image, mode="RGB").resize(new_size, Image.Resampling.BILINEAR)
        return np.asarray(resized), scale
    raise CropError("could not resize image")


def polygon_area(points: np.ndarray) -> float:
    x = points[:, 0]
    y = points[:, 1]
    return float(abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1))) / 2.0)


def order_points(points: np.ndarray) -> np.ndarray:
    ordered = np.zeros((4, 2), dtype=np.float32)
    sums = points.sum(axis=1)
    diffs = np.diff(points, axis=1).reshape(-1)
    ordered[0] = points[np.argmin(sums)]
    ordered[2] = points[np.argmax(sums)]
    ordered[1] = points[np.argmin(diffs)]
    ordered[3] = points[np.argmax(diffs)]
    return ordered


def distance(a: np.ndarray, b: np.ndarray) -> float:
    return float(math.hypot(float(a[0] - b[0]), float(a[1] - b[1])))


def inset_crop(image: np.ndarray, inset_ratio: float) -> np.ndarray:
    height, width = image.shape[:2]
    dx = int(round(width * inset_ratio))
    dy = int(round(height * inset_ratio))
    if dx <= 0 and dy <= 0:
        return image
    if width - 2 * dx < 32 or height - 2 * dy < 32:
        return image
    return image[dy : height - dy, dx : width - dx]


def opencv_perspective_crop(image: np.ndarray, inset_ratio: float) -> CropResult | None:
    if cv2 is None:
        return None

    height, width = image.shape[:2]
    small, scale = resize_rgb(image, 1200)
    gray = cv2.cvtColor(small, cv2.COLOR_RGB2GRAY)
    gray = cv2.GaussianBlur(gray, (5, 5), 0)
    edges = cv2.Canny(gray, 35, 120)
    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (5, 5))
    edges = cv2.dilate(edges, kernel, iterations=1)

    contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    candidates: list[tuple[float, np.ndarray, float]] = []
    small_area = float(small.shape[0] * small.shape[1])

    for contour in contours:
        area_ratio = cv2.contourArea(contour) / small_area
        if area_ratio < 0.18 or area_ratio > 0.98:
            continue

        perimeter = cv2.arcLength(contour, True)
        approx = cv2.approxPolyDP(contour, 0.02 * perimeter, True)
        if len(approx) != 4:
            continue

        points = order_points(approx.reshape(4, 2).astype(np.float32))
        top_width = distance(points[0], points[1])
        bottom_width = distance(points[3], points[2])
        left_height = distance(points[0], points[3])
        right_height = distance(points[1], points[2])
        average_width = (top_width + bottom_width) / 2.0
        average_height = (left_height + right_height) / 2.0
        if average_width < 120 or average_height < 120:
            continue

        aspect = average_width / average_height if average_height else 0
        aspect_score = max(0.2, 1.0 - abs(aspect - 0.72))
        score = area_ratio * aspect_score
        candidates.append((score, points / scale, area_ratio))

    if not candidates:
        return None

    _score, source_points, _small_area_ratio = max(candidates, key=lambda item: item[0])
    top_width = distance(source_points[0], source_points[1])
    bottom_width = distance(source_points[3], source_points[2])
    left_height = distance(source_points[0], source_points[3])
    right_height = distance(source_points[1], source_points[2])
    output_width = max(1, int(round(max(top_width, bottom_width))))
    output_height = max(1, int(round(max(left_height, right_height))))

    destination = np.array(
        [
            [0, 0],
            [output_width - 1, 0],
            [output_width - 1, output_height - 1],
            [0, output_height - 1],
        ],
        dtype=np.float32,
    )
    transform = cv2.getPerspectiveTransform(source_points.astype(np.float32), destination)
    warped = cv2.warpPerspective(
        image,
        transform,
        (output_width, output_height),
        flags=cv2.INTER_CUBIC,
        borderMode=cv2.BORDER_REPLICATE,
    )
    warped = inset_crop(warped, inset_ratio)
    area_ratio = polygon_area(source_points) / float(width * height)

    return CropResult(
        image=warped,
        method="opencv-rectangle",
        area_ratio=area_ratio,
        points=[tuple(map(float, point)) for point in source_points],
        confidence=None,
    )


def smooth_scores(scores: np.ndarray, radius: int) -> np.ndarray:
    if radius <= 0:
        return scores
    kernel = np.ones((radius * 2 + 1,), dtype=np.float64)
    numerator = np.convolve(scores, kernel, mode="same")
    denominator = np.convolve(np.ones_like(scores), kernel, mode="same")
    return numerator / denominator


def best_interval(scores: np.ndarray, threshold: float, minimum_length: int) -> tuple[int, int] | None:
    best: tuple[int, int] | None = None
    best_weight = 0.0
    start: int | None = None
    total = 0.0

    for index in range(len(scores) + 1):
        active = index < len(scores) and scores[index] >= threshold
        if active:
            if start is None:
                start = index
                total = 0.0
            total += float(scores[index])
            continue

        if start is None:
            continue

        length = index - start
        if length >= minimum_length and total > best_weight:
            best = (start, index)
            best_weight = total
        start = None
        total = 0.0

    return best


def background_crop(image: np.ndarray) -> CropResult | None:
    height, width = image.shape[:2]
    small, scale = resize_rgb(image, 900)
    small_height, small_width = small.shape[:2]
    as_float = small.astype(np.float32) / 255.0
    red = as_float[:, :, 0]
    green = as_float[:, :, 1]
    blue = as_float[:, :, 2]
    brightness = 0.2126 * red + 0.7152 * green + 0.0722 * blue
    max_channel = np.maximum.reduce([red, green, blue])
    min_channel = np.minimum.reduce([red, green, blue])
    saturation = np.divide(
        max_channel - min_channel,
        np.maximum(max_channel, 1e-6),
        out=np.zeros_like(max_channel),
    )

    page_like = (brightness > 0.34) & (saturation < 0.28)
    column_scores = smooth_scores(page_like.mean(axis=0).astype(np.float64), max(3, small_width // 100))
    row_scores = smooth_scores(page_like.mean(axis=1).astype(np.float64), max(3, small_height // 100))

    x_range = best_interval(column_scores, threshold=0.42, minimum_length=small_width // 2)
    y_range = best_interval(row_scores, threshold=0.42, minimum_length=small_height // 2)
    if x_range is None or y_range is None:
        return None

    x0_small, x1_small = x_range
    y0_small, y1_small = y_range
    x0 = x0_small / scale
    x1 = x1_small / scale
    y0 = y0_small / scale
    y1 = y1_small / scale

    left_padding = width * (0.06 if x0_small > small_width / 20 else 0.012)
    right_padding = width * 0.012
    top_padding = height * 0.012
    bottom_padding = height * 0.06

    left = max(0, int(math.floor(x0 - left_padding)))
    right = min(width, int(math.ceil(x1 + right_padding)))
    top = max(0, int(math.floor(y0 - top_padding)))
    bottom = min(height, int(math.ceil(y1 + bottom_padding)))
    left, right = trim_bright_edge_slivers(image, left, right, top, bottom)
    if right - left < 64 or bottom - top < 64:
        return None

    area_ratio = ((right - left) * (bottom - top)) / float(width * height)
    if area_ratio < 0.18 or area_ratio > 0.995:
        return None

    cropped = image[top:bottom, left:right]
    points = [
        (float(left), float(top)),
        (float(right), float(top)),
        (float(right), float(bottom)),
        (float(left), float(bottom)),
    ]
    return CropResult(
        image=cropped,
        method="background-fallback",
        area_ratio=area_ratio,
        points=points,
        confidence=None,
    )


def trim_bright_edge_slivers(
    image: np.ndarray,
    left: int,
    right: int,
    top: int,
    bottom: int,
) -> tuple[int, int]:
    """Trim narrow bright facing-page strips that remain attached to a page band."""

    crop = image[top:bottom, left:right].astype(np.float32) / 255.0
    height, width = crop.shape[:2]
    if width < 400 or height < 400:
        return left, right

    red = crop[:, :, 0]
    green = crop[:, :, 1]
    blue = crop[:, :, 2]
    brightness = 0.2126 * red + 0.7152 * green + 0.0722 * blue
    dark = brightness < 0.36
    y0 = int(height * 0.08)
    y1 = int(height * 0.92)
    usable_brightness = brightness[y0:y1, :]
    usable_dark = dark[y0:y1, :]

    edge_width = max(24, int(width * 0.035))
    search_width = min(int(width * 0.24), 720)
    minimum_remaining = int(width * 0.65)

    def trim_left(current_left: int) -> int:
        edge_brightness = float(usable_brightness[:, :edge_width].mean())
        edge_dark_ratio = float(usable_dark[:, :edge_width].mean())
        if edge_brightness < 0.62 or edge_dark_ratio < 0.015:
            return current_left

        column_scores = smooth_scores(usable_brightness.mean(axis=0), max(3, width // 180))
        threshold = edge_brightness - 0.14
        candidates = np.flatnonzero(column_scores[:search_width] < threshold)
        if len(candidates) == 0:
            return current_left

        trim = int(candidates[0])
        if trim < int(edge_width * 0.6) or width - trim < minimum_remaining:
            return current_left
        return current_left + trim

    def trim_right(current_right: int) -> int:
        edge_brightness = float(usable_brightness[:, -edge_width:].mean())
        edge_dark_ratio = float(usable_dark[:, -edge_width:].mean())
        if edge_brightness < 0.62 or edge_dark_ratio < 0.015:
            return current_right

        column_scores = smooth_scores(usable_brightness.mean(axis=0), max(3, width // 180))
        threshold = edge_brightness - 0.14
        right_scores = column_scores[width - search_width :]
        candidates = np.flatnonzero(right_scores < threshold)
        if len(candidates) == 0:
            return current_right

        trim = int(search_width - candidates[-1])
        if trim < int(edge_width * 0.6) or width - trim < minimum_remaining:
            return current_right
        return current_right - trim

    return trim_left(left), trim_right(right)


def crop_page(input_path: Path, output_path: Path, inset_ratio: float) -> tuple[CropResult, int, int]:
    image = load_rgb(input_path)
    height, width = image.shape[:2]

    result = opencv_perspective_crop(image, inset_ratio)
    if result is None:
        result = background_crop(image)
    if result is None:
        raise CropError("no page boundary detected")

    save_png(result.image, output_path)
    return result, width, height


def write_report(path: Path | None, report: dict[str, Any]) -> None:
    text = json.dumps(report, ensure_ascii=False, sort_keys=True)
    if path is None:
        print(text)
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text + "\n", encoding="utf-8")


def image_size(path: Path) -> tuple[int, int]:
    try:
        image = load_rgb(path)
    except Exception:
        return (0, 0)
    height, width = image.shape[:2]
    return (width, height)


def build_report(
    *,
    input_path: Path,
    output_path: Path,
    did_crop: bool,
    method: str,
    source_width: int,
    source_height: int,
    output_width: int = 0,
    output_height: int = 0,
    confidence: float | None = None,
    area_ratio: float | None = None,
    points: list[tuple[float, float]] | None = None,
    error: str | None = None,
) -> dict[str, Any]:
    points = points or []
    point_values = [point_report(point) for point in points]
    while len(point_values) < 4:
        point_values.append(None)

    return {
        "areaRatio": area_ratio,
        "bottomLeft": point_values[3],
        "bottomRight": point_values[2],
        "confidence": confidence,
        "didCrop": did_crop,
        "error": error,
        "inputPath": str(input_path),
        "method": method,
        "outputHeight": output_height,
        "outputPath": str(output_path),
        "outputWidth": output_width,
        "sourceHeight": source_height,
        "sourceWidth": source_width,
        "topLeft": point_values[0],
        "topRight": point_values[1],
    }


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Crop a cookbook page image before OCR.")
    parser.add_argument("--metadata", type=Path)
    parser.add_argument("--inset", type=float, default=0.006)
    parser.add_argument("input_image", type=Path)
    parser.add_argument("output_png", type=Path)
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    try:
        result, source_width, source_height = crop_page(args.input_image, args.output_png, args.inset)
        output_height, output_width = result.image.shape[:2]
        report = build_report(
            input_path=args.input_image,
            output_path=args.output_png,
            did_crop=True,
            method=result.method,
            source_width=source_width,
            source_height=source_height,
            output_width=output_width,
            output_height=output_height,
            confidence=result.confidence,
            area_ratio=result.area_ratio,
            points=result.points,
        )
        write_report(args.metadata, report)
        return 0
    except Exception as exc:
        source_width, source_height = image_size(args.input_image)
        report = build_report(
            input_path=args.input_image,
            output_path=args.output_png,
            did_crop=False,
            method="none",
            source_width=source_width,
            source_height=source_height,
            error=str(exc),
        )
        try:
            write_report(args.metadata, report)
        except Exception:
            pass
        print(str(exc), file=sys.stderr)
        return 4


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: tools/ocr/page_crop.swift
import CoreGraphics
import CoreImage
import Foundation
import ImageIO
import UniformTypeIdentifiers
import Vision

struct PointReport: Encodable {
    let x: Double
    let y: Double
}

struct CropReport: Encodable {
    let inputPath: String
    let outputPath: String
    let didCrop: Bool
    let method: String
    let sourceWidth: Int
    let sourceHeight: Int
    let outputWidth: Int
    let outputHeight: Int
    let confidence: Float?
    let areaRatio: Double?
    let topLeft: PointReport?
    let topRight: PointReport?
    let bottomRight: PointReport?
    let bottomLeft: PointReport?
    let error: String?
}

enum CropError: Error, CustomStringConvertible {
    case couldNotLoadImage(String)
    case noRectangleDetected
    case couldNotRenderImage
    case couldNotCreateBitmap
    case couldNotWriteImage(String)

    var description: String {
        switch self {
        case .couldNotLoadImage(let path):
            return "could not load image: \(path)"
        case .noRectangleDetected:
            return "no page rectangle detected"
        case .couldNotRenderImage:
            return "could not render corrected image"
        case .couldNotCreateBitmap:
            return "could not create analysis bitmap"
        case .couldNotWriteImage(let path):
            return "could not write image: \(path)"
        }
    }
}

func writeJson<T: Encodable>(_ value: T, to path: String?) throws {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    let data = try encoder.encode(value)

    if let path {
        try data.write(to: URL(fileURLWithPath: path))
    } else {
        FileHandle.standardOutput.write(data)
        FileHandle.standardOutput.write(Data([0x0a]))
    }
}

func imagePixelSize(_ path: String) -> (Int, Int) {
    let url = URL(fileURLWithPath: path)
    guard
        let source = CGImageSourceCreateWithURL(url as CFURL, nil),
        let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any]
    else {
        return (0, 0)
    }

    let width = properties[kCGImagePropertyPixelWidth] as? Int ?? 0
    let height = properties[kCGImagePropertyPixelHeight] as? Int ?? 0
    return (width, height)
}

func point(_ normalized: CGPoint, width: Double, height: Double) -> CGPoint {
    CGPoint(x: normalized.x * width, y: normalized.y * height)
}

func reportPoint(_ value: CGPoint) -> PointReport {
    PointReport(x: value.x, y: value.y)
}

func polygonArea(_ points: [CGPoint]) -> Double {
    guard points.count > 2 else {
        return 0
    }

    var total = 0.0
    for index in points.indices {
        let next = points[(index + 1) % points.count]
        total += points[index].x * next.y
        total -= next.x * points[index].y
    }

    return abs(total) / 2
}

func distance(_ a: CGPoint, _ b: CGPoint) -> Double {
    hypot(a.x - b.x, a.y - b.y)
}

func rectangleScore(_ observation: VNRectangleObservation, width: Double, height: Double) -> Double {
    let topLeft = point(observation.topLeft, width: width, height: height)
    let topRight = point(observation.topRight, width: width, height: height)
    let bottomRight = point(observation.bottomRight, width: width, height: height)
    let bottomLeft = point(observation.bottomLeft, width: width, height: height)
    let area = polygonArea([topLeft, topRight, bottomRight, bottomLeft])
    let areaRatio = area / (width * height)

    let topWidth = distance(topLeft, topRight)
    let bottomWidth = distance(bottomLeft, bottomRight)
    let leftHeight = distance(topLeft, bottomLeft)
    let rightHeight = distance(topRight, bottomRight)
    let averageWidth = (topWidth + bottomWidth) / 2
    let averageHeight = (leftHeight + rightHeight) / 2
    let aspectRatio = averageHeight > 0 ? averageWidth / averageHeight : 0

    let aspectScore = max(0.2, 1 - abs(aspectRatio - 0.72))
    return Double(observation.confidence) * areaRatio * aspectScore
}

struct AxisAlignedCrop {
    let rect: CGRect
    let areaRatio: Double
}

struct AnalysisBitmap {
    let width: Int
    let height: Int
    let pageLike: [Bool]
}

func bestRectangle(_ observations: [VNRectangleObservation], width: Double, height: Double) -> VNRectangleObservation? {
    observations
        .filter { observation in
            let points = [
                point(observation.topLeft, width: width, height: height),
                point(observation.topRight, width: width, height: height),
                point(observation.bottomRight, width: width, height: height),
                point(observation.bottomLeft, width: width, height: height),
            ]
            let areaRatio = polygonArea(points) / (width * height)
            return areaRatio >= 0.18 && areaRatio <= 0.98
        }
        .max { first, second in
            rectangleScore(first, width: width, height: height)
                < rectangleScore(second, width: width, height: height)
        }
}

func smoothed(_ scores: [Double], radius: Int) -> [Double] {
    guard radius > 0, !scores.isEmpty else {
        return scores
    }

    var prefix = [Double](repeating: 0, count: scores.count + 1)
    for index in scores.indices {
        prefix[index + 1] = prefix[index] + scores[index]
    }

    return scores.indices.map { index in
        let start = max(0, index - radius)
        let end = min(scores.count, index + radius + 1)
        return (prefix[end] - prefix[start]) / Double(end - start)
    }
}

func bestInterval(_ scores: [Double], threshold: Double, minimumLength: Int) -> Range<Int>? {
    var best: Range<Int>?
    var bestWeight = 0.0
    var start: Int?
    var total = 0.0

    for index in 0...scores.count {
        let active = index < scores.count && scores[index] >= threshold
        if active {
            if start == nil {
                start = index
                total = 0
            }
            total += scores[index]
            continue
        }

        guard let intervalStart = start else {
            continue
        }

        let length = index - intervalStart
        if length >= minimumLength, total > bestWeight {
            best = intervalStart..<index
            bestWeight = total
        }
        start = nil
        total = 0
    }

    return best
}

func analysisBitmap(from image: CIImage, maxDimension: CGFloat = 900) throws -> AnalysisBitmap {
    let extent = image.extent.integral
    let scale = min(1, maxDimension / max(extent.width, extent.height))
    let analysisImage = image.transformed(by: CGAffineTransform(scaleX: scale, y: scale))
    let context = CIContext()
    guard let cgImage = context.createCGImage(analysisImage, from: analysisImage.extent.integral) else {
        throw CropError.couldNotRenderImage
    }

    let width = cgImage.width
    let height = cgImage.height
    let bytesPerPixel = 4
    let bytesPerRow = width * bytesPerPixel
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    var pixels = [UInt8](repeating: 0, count: height * bytesPerRow)

    try pixels.withUnsafeMutableBytes { rawBuffer in
        guard let bitmap = CGContext(
            data: rawBuffer.baseAddress,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: bytesPerRow,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else {
            throw CropError.couldNotCreateBitmap
        }
        bitmap.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    }

    var pageLike = [Bool](repeating: false, count: width * height)
    for y in 0..<height {
        for x in 0..<width {
            let offset = y * bytesPerRow + x * bytesPerPixel
            let red = Double(pixels[offset])
            let green = Double(pixels[offset + 1])
            let blue = Double(pixels[offset + 2])
            let maxChannel = max(red, green, blue)
            let minChannel = min(red, green, blue)
            let brightness = (0.2126 * red + 0.7152 * green + 0.0722 * blue) / 255
            let saturation = maxChannel > 0 ? (maxChannel - minChannel) / maxChannel : 0

            pageLike[y * width + x] = brightness > 0.34 && saturation < 0.28
        }
    }

    return AnalysisBitmap(width: width, height: height, pageLike: pageLike)
}

func fallbackAxisAlignedCrop(_ image: CIImage) throws -> AxisAlignedCrop? {
    let bitmap = try analysisBitmap(from: image)
    let width = bitmap.width
    let height = bitmap.height

    var columnScores = [Double](repeating: 0, count: width)
    var rowScores = [Double](repeating: 0, count: height)

    for y in 0..<height {
        for x in 0..<width where bitmap.pageLike[y * width + x] {
            columnScores[x] += 1
            rowScores[y] += 1
        }
    }

    columnScores = smoothed(columnScores.map { $0 / Double(height) }, radius: max(3, width / 100))
    rowScores = smoothed(rowScores.map { $0 / Double(width) }, radius: max(3, height / 100))

    guard
        let xRange = bestInterval(columnScores, threshold: 0.42, minimumLength: width / 2),
        let yRange = bestInterval(rowScores, threshold: 0.42, minimumLength: height / 2)
    else {
        return nil
    }

    let scaleX = image.extent.width / Double(width)
    let scaleY = image.extent.height / Double(height)
    let leftPadding = xRange.lowerBound > width / 20 ? image.extent.width * 0.06 : image.extent.width * 0.012
    let rightPadding = image.extent.width * 0.012
    let topPadding = image.extent.height * 0.012
    let bottomPadding = image.extent.height * 0.06

    let minX = max(image.extent.minX, Double(xRange.lowerBound) * scaleX - leftPadding)
    let maxX = min(image.extent.maxX, Double(xRange.upperBound) * scaleX + rightPadding)
    let topY = max(0, Double(yRange.lowerBound) * scaleY - topPadding)
    let bottomY = min(image.extent.height, Double(yRange.upperBound) * scaleY + bottomPadding)
    let minY = max(image.extent.minY, image.extent.maxY - bottomY)
    let maxY = min(image.extent.maxY, image.extent.maxY - topY)
    let rect = CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
    let areaRatio = (rect.width * rect.height) / (image.extent.width * image.extent.height)

    guard areaRatio >= 0.18, areaRatio <= 0.99 else {
        return nil
    }

    return AxisAlignedCrop(rect: rect, areaRatio: areaRatio)
}

func writePng(_ image: CIImage, to path: String) throws -> (Int, Int) {
    let context = CIContext()
    let extent = image.extent.integral
    guard let cgImage = context.createCGImage(image, from: extent) else {
        throw CropError.couldNotRenderImage
    }

    let url = URL(fileURLWithPath: path)
    guard let destination = CGImageDestinationCreateWithURL(url as CFURL, UTType.png.identifier as CFString, 1, nil) else {
        throw CropError.couldNotWriteImage(path)
    }
    CGImageDestinationAddImage(destination, cgImage, nil)
    guard CGImageDestinationFinalize(destination) else {
        throw CropError.couldNotWriteImage(path)
    }

    return (cgImage.width, cgImage.height)
}

func cropPage(inputPath: String, outputPath: String, insetRatio: Double, metadataPath: String?) throws {
    let inputUrl = URL(fileURLWithPath: inputPath)
    guard let sourceImage = CIImage(contentsOf: inputUrl, options: [.applyOrientationProperty: true]) else {
        throw CropError.couldNotLoadImage(inputPath)
    }

    let extent = sourceImage.extent
    let width = extent.width
    let height = extent.height

    let request = VNDetectRectanglesRequest()
    request.maximumObservations = 12
    request.minimumConfidence = 0.35
    request.minimumSize = 0.18
    request.minimumAspectRatio = 0.25
    request.maximumAspectRatio = 1.15
    request.quadratureTolerance = 35

    let handler = VNImageRequestHandler(ciImage: sourceImage, options: [:])
    try handler.perform([request])

    if let rectangle = bestRectangle(request.results ?? [], width: width, height: height) {
        let topLeft = point(rectangle.topLeft, width: width, height: height)
        let topRight = point(rectangle.topRight, width: width, height: height)
        let bottomRight = point(rectangle.bottomRight, width: width, height: height)
        let bottomLeft = point(rectangle.bottomLeft, width: width, height: height)

        let corrected = sourceImage
            .applyingFilter(
                "CIPerspectiveCorrection",
                parameters: [
                    "inputTopLeft": CIVector(cgPoint: topLeft),
                    "inputTopRight": CIVector(cgPoint: topRight),
                    "inputBottomRight": CIVector(cgPoint: bottomRight),
                    "inputBottomLeft": CIVector(cgPoint: bottomLeft),
                ]
            )

        let correctedExtent = corrected.extent
        let dx = correctedExtent.width * insetRatio
        let dy = correctedExtent.height * insetRatio
        let finalImage = corrected.cropped(to: correctedExtent.insetBy(dx: dx, dy: dy))
        let outputSize = try writePng(finalImage, to: outputPath)

        let areaRatio = polygonArea([topLeft, topRight, bottomRight, bottomLeft]) / (width * height)
        try writeJson(
            CropReport(
                inputPath: inputPath,
                outputPath: outputPath,
                didCrop: true,
                method: "vision-rectangle",
                sourceWidth: Int(width.rounded()),
                sourceHeight: Int(height.rounded()),
                outputWidth: outputSize.0,
                outputHeight: outputSize.1,
                confidence: rectangle.confidence,
                areaRatio: areaRatio,
                topLeft: reportPoint(topLeft),
                topRight: reportPoint(topRight),
                bottomRight: reportPoint(bottomRight),
                bottomLeft: reportPoint(bottomLeft),
                error: nil
            ),
            to: metadataPath
        )
        return
    }

    guard let fallback = try fallbackAxisAlignedCrop(sourceImage) else {
        throw CropError.noRectangleDetected
    }

    let finalImage = sourceImage.cropped(to: fallback.rect)
    let outputSize = try writePng(finalImage, to: outputPath)
    let topLeft = CGPoint(x: fallback.rect.minX, y: fallback.rect.maxY)
    let topRight = CGPoint(x: fallback.rect.maxX, y: fallback.rect.maxY)
    let bottomRight = CGPoint(x: fallback.rect.maxX, y: fallback.rect.minY)
    let bottomLeft = CGPoint(x: fallback.rect.minX, y: fallback.rect.minY)

    try writeJson(
        CropReport(
            inputPath: inputPath,
            outputPath: outputPath,
            didCrop: true,
            method: "brightness-fallback",
            sourceWidth: Int(width.rounded()),
            sourceHeight: Int(height.rounded()),
            outputWidth: outputSize.0,
            outputHeight: outputSize.1,
            confidence: nil,
            areaRatio: fallback.areaRatio,
            topLeft: reportPoint(topLeft),
            topRight: reportPoint(topRight),
            bottomRight: reportPoint(bottomRight),
            bottomLeft: reportPoint(bottomLeft),
            error: nil
        ),
        to: metadataPath
    )
}

let arguments = Array(CommandLine.arguments.dropFirst())
if arguments.isEmpty {
    FileHandle.standardError.write(
        Data("usage: swift tools/ocr/page_crop.swift [--metadata <path>] [--inset <ratio>] <input-image> <output-png>\n".utf8)
    )
    exit(2)
}

var metadataPath: String?
var insetRatio = 0.006
var paths: [String] = []
var index = 0

while index < arguments.count {
    let argument = arguments[index]

    if argument == "--metadata" {
        guard index + 1 < arguments.count else {
            FileHandle.standardError.write(Data("missing value for --metadata\n".utf8))
            exit(2)
        }
        metadataPath = arguments[index + 1]
        index += 2
        continue
    }

    if argument == "--inset" {
        guard index + 1 < arguments.count, let value = Double(arguments[index + 1]) else {
            FileHandle.standardError.write(Data("invalid value for --inset\n".utf8))
            exit(2)
        }
        insetRatio = value
        index += 2
        continue
    }

    paths.append(argument)
    index += 1
}

guard paths.count == 2 else {
    FileHandle.standardError.write(Data("expected input and output paths\n".utf8))
    exit(2)
}

do {
    try cropPage(inputPath: paths[0], outputPath: paths[1], insetRatio: insetRatio, metadataPath: metadataPath)
} catch {
    let sourceSize = imagePixelSize(paths[0])
    let report = CropReport(
        inputPath: paths[0],
        outputPath: paths[1],
        didCrop: false,
        method: "none",
        sourceWidth: sourceSize.0,
        sourceHeight: sourceSize.1,
        outputWidth: 0,
        outputHeight: 0,
        confidence: nil,
        areaRatio: nil,
        topLeft: nil,
        topRight: nil,
        bottomRight: nil,
        bottomLeft: nil,
        error: String(describing: error)
    )
    try? writeJson(report, to: metadataPath)
    FileHandle.standardError.write(Data("\(error)\n".utf8))
    exit(4)
}
